src/url_aggregator.cpp

```diff
@@ -26,6 +29,63 @@ inline void url_aggregator::set_scheme(std::string_view new_scheme) noexcept {
   // TODO: implement
 }
 
+bool url_aggregator::set_protocol(const std::string_view input) {
+  (void) input;
+  // TODO: Implement
+  return false;
+}
+
+bool url_aggregator::set_username(const std::string_view input) {
+  if (cannot_have_credentials_or_port()) { return false; }
+  size_t username_start = components.protocol_end + 3;
+  size_t username_length = components.username_end - username_start;
+  buffer.erase(username_start, components.username_end);
+
+  std::string encoded_input = ada::unicode::percent_encode(input, character_sets::USERINFO_PERCENT_ENCODE);
+  buffer.append(encoded_input, username_start);
```

Comment:
You *may* use 'buffer.insert' here. But then maybe not.

Reply:
Insert seems more natural, I like it.